### PR TITLE
[google_sign_in] Port plugin to use the federated Platform Interface

### PR DESCRIPTION
**Pull request by ditman**
## Description

* Migrate the 'native' plugin to the Platform Interface
* Adapt tests to the new API

This is required for plugin federation (to implement the Web version in a separate package)

**The `GoogleSignInPlatform.instance` interface is defined (and implemented) in the `google_sign_in_platform_interface` package, [here](https://pub.dev/documentation/google_sign_in_platform_interface/latest/google_sign_in_platform_interface/google_sign_in_platform_interface-library.html).**

## Related Issues

https://github.com/flutter/flutter/issues/43775

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 1]
```diff
@@ -157,18 +147,22 @@ class GoogleSignIn {
   /// The [hostedDomain] argument specifies a hosted domain restriction. By
   /// setting this, sign in will be restricted to accounts of the user in the
   /// specified domain. By default, the list of accounts will not be restricted.
-  GoogleSignIn({this.signInOption, this.scopes, this.hostedDomain});
+  GoogleSignIn(
+      {this.signInOption = SignInOption.standard,
+      this.scopes = const <String>[],
+      this.hostedDomain});
```
Comment (hterkelsen) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: nit: this will format a little nicer if you put a `,` after `this.hostedDomain`

Comment (ditman) [reply packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: Thanks! I'll incorporate this change, and reformat, while rebasing the branch with the latest from master!

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 2]
```diff
@@ -146,7 +139,7 @@ class GoogleSignIn {
   /// Initializes global sign-in configuration settings.
   ///
   /// The [signInOption] determines the user experience. [SigninOption.games]
-  /// must not be used on iOS.
+  /// must not be used on iOS or web.
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: Maybe we should just say:
`is only supported on Android.`

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 3]
```diff
@@ -157,18 +150,25 @@ class GoogleSignIn {
...
+  }) {
     return GoogleSignIn(
         signInOption: SignInOption.standard,
         scopes: scopes,
         hostedDomain: hostedDomain);
   }
 
   /// Factory for creating sign in suitable for games. This option must not be
-  /// used on iOS because the games API is not supported.
+  /// used on iOS or web because the games API is not supported.
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: ditto

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 4]
```diff
@@ -186,13 +186,8 @@ class GoogleSignIn {
...
   static const String kSignInFailedError = 'sign_in_failed';
 
-  /// The [MethodChannel] over which this class communicates.
-  @visibleForTesting
-  static const MethodChannel channel =
-      MethodChannel('plugins.flutter.io/google_sign_in');
-
   /// Option to determine the sign in user experience. [SignInOption.games] must
-  /// not be used on iOS.
+  /// not be used on iOS or web.
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: ditto

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 5]
```diff
@@ -211,12 +206,12 @@ class GoogleSignIn {
...
   Future<void> _initialization;
 
-  Future<GoogleSignInAccount> _callMethod(String method) async {
+  Future<GoogleSignInAccount> _callMethod(Function method) async {
     await _ensureInitialized();
 
-    final Map<String, dynamic> response =
-        await channel.invokeMapMethod<String, dynamic>(method);
-    return _setCurrentUser(response != null && response.isNotEmpty
+    final dynamic response = await method();
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: what would happen if the `method` is not an async function.

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 6]
```diff
@@ -251,6 +244,7 @@ class GoogleSignIn {
     final Completer<void> completer = Completer<void>();
     future.whenComplete(completer.complete).catchError((dynamic _) {
       // Ignore if previous call completed with an error.
+      // TODO: Should we log errors here, if debug or similar?
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: Yes. let's log an error here. Personally, I think we should be clear about things like this error is completely harmless and it is logged for debug purpose only.
@mehmetf I saw that this is added in https://github.com/flutter/plugins/commit/4eeb744bfce3cbacd64dcce32b52295c3299fe89
Did you recall why did we ignore this error?

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart [hunk 7]
```diff
@@ -259,26 +253,25 @@ class GoogleSignIn {
   ///
   /// At most one in flight call is allowed to prevent concurrent (out of order)
   /// updates to [currentUser] and [onCurrentUserChanged].
-  Future<GoogleSignInAccount> _addMethodCall(String method) async {
+  Future<GoogleSignInAccount> _addMethodCall(
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: nits(optional): Add doc for the new parameter 'canSkipCall`.

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart (hunk 6 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: I'll revisit this to add logging, so you guys can discuss how much of it we need :)

(I'd recommend that this gets logged at the lowest level that makes sense, like trace or similar)

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart (hunk 7 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: Added!

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart (hunk 5 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: Good call! It seems to work, though, check this [DartPad](https://dartpad.dartlang.org/f062d40311b09b39c6bdef925137653e)!

File: packages/google_sign_in/google_sign_in/lib/google_sign_in.dart (hunk 2 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in/lib/google_sign_in.dart]: Good call, much clearer.

**Pull request merged**